File: Formularios/Agregar/add_stock.cs
using System;
using System.Data;
using System.Linq;
using System.Windows.Forms;

namespace Centrex
{

    public partial class add_stock
    {
        public bool ingreso_guardado;
        private CentrexDbContext ctx = new CentrexDbContext();

        public add_stock()
        {
            InitializeComponent();
        }

        private void add_stock_Load(object sender, EventArgs e)
        {
            // Cargar proveedores con EF
            var proveedores = ctx.ProveedorEntity.Where(p => p.Activo == true).OrderBy(p => p.RazonSocial).ToList();

            cmb_proveedor.DataSource = proveedores;
            cmb_proveedor.DisplayMember = "RazonSocial";
            cmb_proveedor.ValueMember = "IdProveedor";

            if (edicion)
            {
                txt_fecha.Text = DateTime.Parse(Conversions.ToString(edita_registro_stock.Fecha.Value)).ToString("dd/MM/yyyy");
                lbl_fecha_ingreso.Text = edita_registro_stock.FechaIngreso.ToString("dd/MM/yyyy");
                txt_factura.Text = edita_registro_stock.Factura;
                cmb_proveedor.SelectedValue = (byte)edita_registro_stock.IdProveedor;

                if (!editaStock)
                {
                    txt_fecha.Enabled = false;
                    txt_factura.Enabled = false;
                    txt_notas.Enabled = false;
                    cmb_proveedor.Enabled = false;
                    psearch_proveedor.Enabled = false;
                    cmd_additem.Enabled = false;
                }
            }
            else
            {
                txt_fecha.Text = DateTime.Now.ToString("dd/MM/yyyy");
                lbl_fecha_ingreso.Text = DateTime.Now.ToString("dd/MM/yyyy");

                // Valor por defecto si existe un proveedor base
                try
                {
                    cmb_proveedor.SelectedValue = 20;
                }
                catch
                {
                }

                // Limpiar tabla temporal
                var tmp = ctx.TmpRegistroStockEntity.ToList();
                if (tmp.Any())
                {
                    ctx.TmpRegistroStockEntity.RemoveRange(tmp);
                    ctx.SaveChanges();
                }
            }

            updateform();
        }

        private void add_stock_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (ingreso_guardado)
                return;
            if (!edicion)
            {
                if (Interaction.MsgBox("Eliminará el ingreso, por lo cual no se contabilizará. ¿Deseá continuar?", (MsgBoxStyle)((int)Constants.vbYesNo + (int)Constants.vbQuestion)) == Constants.vbNo)
                {
                    e.Cancel = true;
                }
            }
        }

        // ==========================================================
        // Refresca el contenido del DataGridView
        // ==========================================================
        private void updateform()
        {
            object data;

            if (edicion && editaStock)
            {
                data = ctx.RegistroStockEntity.Include("Item").Where(rs => rs.IdIngreso == edita_registro_stock.IdIngreso).Select(rs => new
                {
                    ID = rs.IdRegistro,
                    Código = rs.IdItemNavigation.Item,
                    Producto = rs.IdItemNavigation.Descript,
                    rs.Factura,
                    rs.Fecha,
                    rs.Cantidad,
                    rs.Costo,
                    rs.PrecioLista,
                    rs.Factor,
                    Notas = rs.Nota
                }).ToList();
            }
            else
            {
                data = ctx.TmpRegistroStockEntity.Include("Item").Where(rst => rst.Activo == true).Select(rst => new
                {
                    ID = rst.IdRegistrotmp,
                    Código = rst.IdItemNavigation.Item,
                    Producto = rst.IdItemNavigation.Descript,
                    Factura = rst.Factura,
                    Fecha = rst.Fecha,
                    Cantidad = rst.Cantidad,
                    Costo = rst.Costo,
                    rst.PrecioLista,
                    Factor = rst.Factor,
                    Notas = rst.Nota
                }).ToList();
            }

            dg_view.AutoGenerateColumns = true;
            dg_view.DataSource = data;
        }

        // ==========================================================
        // Guarda el ingreso (trasladando tmp → registros_stock)
        // ==========================================================
        private void cmd_ok_Click(object sender, EventArgs e)
        {
            if (!edicion)
            {
                if (Interaction.MsgBox("Se modificarán todos los items anteriormente cargados. ¿Está seguro de que desea continuar?", (MsgBoxStyle)((int)Constants.vbYesNo + (int)Constants.vbQuestion), "Centrex") == Constants.vbYes)
                {
                    stock.AddStock();
                    // Limpiar tabla temporal
                    var tmp = ctx.TmpRegistroStockEntity.ToList();
                    if (tmp.Any())
                    {
                        ctx.TmpRegistroStockEntity.RemoveRange(tmp);
                        ctx.SaveChanges();
                    }
                    Interaction.MsgBox("Se actualizaron los items correctamente", (MsgBoxStyle)Constants.vbOK, "Centrex");
                    ingreso_guardado = true;
                }
                else
                {
                    return;
                }
            }

            closeandupdate(this);
        }

        // ==========================================================
        // Agregar ítem (abre buscador + formulario de carga)
        // ==========================================================
        private void cmd_additem_Click(object sender, EventArgs e)
        {
            Enabled = false;
            My.MyProject.Forms.search.ShowDialog();
            Enabled = true;

            if (id == 0)
                return;

            var frm = new infoagregarstock(Conversions.ToInteger(cmb_proveedor.SelectedValue));
            frm.ShowDialog();
            updateform();
        }

        // ==========================================================
        // Cancelar ingreso (cierra sin guardar)
        // ==========================================================
        private void cmd_cancel_Click(object sender, EventArgs e)
        {
            if (!edicion)
            {
                if (Interaction.MsgBox("Eliminará el ingreso, por lo cual no se contabilizará. ¿Deseá continuar?", (MsgBoxStyle)((int)Constants.vbYesNo + (int)Constants.vbQuestion)) == Constants.vbYes)
                {
                    ingreso_guardado = true;
                    closeandupdate(this);
                }
            }
            else
            {
                closeandupdate(this);
            }
        }

        // ==========================================================
        // Buscar proveedor
        // ==========================================================
        private void psearch_proveedor_Click(object sender, EventArgs e)
        {
            string tmpTabla = tabla;
            tabla = "proveedores";
            Enabled = false;
            My.MyProject.Forms.search.ShowDialog();
            Enabled = true;
            tabla = tmpTabla;

            var proveedor = ctx.ProveedorEntity.FirstOrDefault(p => p.IdProveedor == id);
            if (proveedor is not null)
            {
                cmb_proveedor.SelectedIndex = cmb_proveedor.FindString(proveedor.RazonSocial);
            }
            id = 0;
        }

        // ==========================================================
        // Editar registro (doble clic)
        // ==========================================================
        private void dg_view_CellMouseDoubleClick(object sender, DataGridViewCellMouseEventArgs e)
        {
            if (edicion && !editaStock)
                return;
            if (dg_view.Rows.Count == 0)
                return;

            int seleccionado = Conversions.ToInteger(dg_view.CurrentRow.Cells[0].Value);

            edicion_item_registro_stock = true;
            edita_item_registro_stock = ctx.RegistroStockEntity.FirstOrDefault(r => r.IdRegistro == seleccionado);
            if (edita_item_registro_stock is not null)
            {
                My.MyProject.Forms.infoagregarstock.ShowDialog();
                updateform();
            }
            edicion_item_registro_stock = false;
        }

        // ==========================================================
        // Borrar item del grid temporal
        // ==========================================================
        private void BorrarToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (dg_view.Rows.Count == 0)
                return;

            int seleccionado = Conversions.ToInteger(dg_view.CurrentRow.Cells[0].Value);
            var tmpItem = ctx.TmpRegistroStockEntity.FirstOrDefault(t => t.IdRegistrotmp == seleccionado);

            if (tmpItem is not null)
            {
                var item = ctx.ItemEntity.FirstOrDefault(i => i.IdItem == tmpItem.IdItem);
                string nombreItem = item is not null ? item.Item : "(Desconocido)";

                if (Interaction.MsgBox("¿Está seguro de borrar el item: " + nombreItem + "?", (MsgBoxStyle)((int)Constants.vbYesNo + (int)Constants.vbQuestion)) == Constants.vbYes)
                {
                    ctx.TmpRegistroStockEntity.Remove(tmpItem);
                    ctx.SaveChanges();
                    updateform();
                }
            }
        }

        // ==========================================================
        // Eventos UI menores
        // ==========================================================
        private void EditarToolStripMenuItem_Click(object sender, EventArgs e)
        {
            dg_view_CellMouseDoubleClick(null, null);
        }

        private void lsv_items_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Right)
            {
                if (edicion)
                    ContextMenuStrip1.Enabled = false;
            }
        }

        private void lbl_fecha_ingreso_MouseEnter(object sender, EventArgs e)
        {
            lbl_tooltip.Visible = true;
        }

        private void lbl_fecha_ingreso_MouseLeave(object sender, EventArgs e)
        {
            lbl_tooltip.Visible = false;
        }

        private void cmb_proveedor_KeyPress(object sender, KeyPressEventArgs e)
        {
            e.Handled = true;
        }
    }
}

File: __Viejo/Formularios/Agregar/CS/add_stock.vb
Public Class add_stock
    Public ingreso_guardado As Boolean

    Private Sub add_stock_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        form = Me
        cargar_combo(cmb_proveedor, "SELECT id_proveedor, razon_social FROM proveedores ORDER BY razon_social ASC", basedb, "razon_social", "id_proveedor")
        If edicion Then
            txt_fecha.Text = DateTime.Parse(edita_registro_stock.fecha)
            lbl_fecha_ingreso.Text = DateTime.Parse(edita_registro_stock.fecha_ingreso)
            txt_factura.Text = edita_registro_stock.factura
            cmb_proveedor.SelectedValue = CByte(edita_registro_stock.id_proveedor)
            If Not editaStock Then
                txt_fecha.Enabled = False
                txt_factura.Enabled = False
                txt_notas.Enabled = False
                cmb_proveedor.Enabled = False
                psearch_proveedor.Enabled = False
                cmd_additem.Enabled = False
            End If
        Else
            txt_fecha.Text = Format(DateTime.Now, "dd/MM/yyyy")
            lbl_fecha_ingreso.Text = Format(DateTime.Now, "dd/MM/yyyy")
            cmb_proveedor.SelectedValue = CByte(20)
            borrartbl("tmpregistros_stock")
        End If
        updateform()
    End Sub

    Private Sub add_stock_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        If ingreso_guardado Then Exit Sub
        If edicion = False Then
            If MsgBox("Eliminará el ingreso, por lo cual no se contabilizará. ¿Deseá continuar?", vbYesNo + vbQuestion) = vbNo Then
                e.Cancel = True
            End If
        End If
    End Sub
    Private Sub updateform()
        Dim sqlstr As String
        If edicion And editaStock Then
            sqlstr = "SELECT rs.id_registro AS 'ID', i.item AS 'Código', i.descript AS 'Producto', rs.factura AS 'Factura', rs.fecha AS 'Fecha', rs.cantidad AS 'Cantidad', rs.costo AS 'Costo', " &
                        "rs.precio_lista AS 'Precio de lista', rs.factor AS 'Factor', rs.nota AS 'Notas' " &
                        "FROM registros_stock AS rs " &
                        "LEFT JOIN items AS i ON rs.id_item = i.id_item " &
                        "WHERE rs.id_ingreso = '" + edita_registro_stock.id_ingreso.ToString + "' " &
                        "ORDER BY rs.id_ingreso ASC"
        Else
            sqlstr = "SELECT rstmp.id_registrotmp AS 'ID', i.item AS 'Código', i.descript AS 'Producto', rstmp.factura AS 'Factura', rstmp.fecha AS 'Fecha', rstmp.cantidad AS 'Cantidad', rstmp.costo AS 'Costo', " &
                    "rstmp.precio_lista AS 'Precio de lista', rstmp.factor AS 'Factor', rstmp.nota AS 'Notas' " &
                    "FROM tmpregistros_stock AS rstmp " &
                    "LEFT JOIN items AS i ON rstmp.id_item = i.id_item " &
                    "WHERE rstmp.activo = '1' " &
                    "ORDER BY rstmp.id_ingreso ASC"
        End If


        cargar_datagrid(dg_view, sqlstr, basedb)
        'resaltarColumna(dg_view, 7, Color.Red)
    End Sub

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        If edicion = False Then
            If MsgBox("Se modificarán todos los items anteriormente cargados. ¿Está seguro de que desea continuar?", vbYesNo + vbQuestion, "Centrex") = vbYes Then
                addstock()
                borrartbl("tmpregistros_stock")
                MsgBox("Se actualizaron los items correctamente", vbOK, "Centrex")
                ingreso_guardado = True
            Else
                'MsgBox("Operación cancelada.", vbOKOnly + vbInformation, "Centrex")
                Exit Sub
            End If
        End If

        closeandupdate(Me)
    End Sub
    Private Sub cmd_additem_Click(sender As Object, e As EventArgs) Handles cmd_additem.Click
        Me.Enabled = False
        search.ShowDialog()

        If id = 0 Then Exit Sub
        Dim frm As New infoagregarstock(cmb_proveedor.SelectedValue)
        frm.ShowDialog()
        'infoagregarstock.ShowDialog()
        updateform()
        'If secuencia Then cmd_additem_Click(Nothing, Nothing)
    End Sub

    Private Sub cmd_cancel_Click(sender As Object, e As EventArgs) Handles cmd_cancel.Click
        If edicion = False Then
            If MsgBox("Eliminará el ingreso, por lo cual no se contabilizará. ¿Deseá continuar?", vbYesNo + vbQuestion) = vbYes Then
                ingreso_guardado = True
                closeandupdate(Me)
            End If
        Else
            closeandupdate(Me)
        End If
    End Sub

    Private Sub psearch_proveedor_Click(sender As Object, e As EventArgs) Handles psearch_proveedor.Click
        Dim tmp As String
        tmp = tabla
        tabla = "proveedores"
        Me.Enabled = False
        search.ShowDialog()
        tabla = tmp

        'Establezco la opción del combo
        cmb_proveedor.SelectedIndex = cmb_proveedor.FindString(info_proveedor(id).razon_social)
        id = 0
    End Sub

    Private Sub EditarToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles EditarToolStripMenuItem.Click
        dg_view_CellMouseDoubleClick(Nothing, Nothing)
    End Sub

    Private Sub lsv_items_MouseDown(sender As Object, e As MouseEventArgs)
        If e.Button = Windows.Forms.MouseButtons.Right Then
            If edicion = True Then
                ContextMenuStrip1.Enabled = False
            End If
        End If
    End Sub

    Private Sub BorrarToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles BorrarToolStripMenuItem.Click
        'borrar item de tmp
        If dg_view.Rows.Count = 0 Then Exit Sub

        Dim seleccionado As String = dg_view.CurrentRow.Cells(0).Value.ToString
        If MsgBox("¿Está seguro de borrar el item: " + info_item(info_registro_stocktmp(seleccionado).id_item).item + "?", vbYesNo + vbQuestion) = vbYes Then
            borraritemregistrostocktmp(info_registro_stocktmp(seleccionado))
        End If
        updateform()
    End Sub

    Private Sub lbl_fecha_ingreso_MouseEnter(sender As Object, e As EventArgs) Handles lbl_fecha_ingreso.MouseEnter
        lbl_tooltip.Visible = True
    End Sub

    Private Sub lbl_fecha_ingreso_MouseLeave(sender As Object, e As EventArgs) Handles lbl_fecha_ingreso.MouseLeave
        lbl_tooltip.Visible = False
    End Sub

    Private Sub dg_view_CellMouseDoubleClick(sender As Object, e As DataGridViewCellMouseEventArgs) Handles dg_view.CellMouseDoubleClick
        If edicion And Not editaStock Then Exit Sub
        If dg_view.Rows.Count = 0 Then Exit Sub

        Dim seleccionado As String = dg_view.CurrentRow.Cells(0).Value.ToString

        edicion_item_registro_stock = True
        edita_item_registro_stock = info_registro_stock(seleccionado)
        infoagregarstock.ShowDialog()
        updateform()
        edicion_item_registro_stock = False
    End Sub

    Private Sub cmb_proveedor_KeyPress(sender As Object, e As KeyPressEventArgs) Handles cmb_proveedor.KeyPress
        'e.KeyChar = ""
    End Sub
End Class